### DISCUSSION TITLE
Running from RAM, build/zephyr/zephyr.bin missing reset vector

### DISCUSSION THREAD

**Discussion by asteriskSF** [zephyrproject-rtos/zephyr#69630] [Q&A] [answered]
I'm working on a project based on NXP MIMXRT106x processor.  The design does not include any flash and the intention is to download the image over USB.

I have found some discussion threads that cover this topic here on zephyr forum as well as parts on the NXP forums.

I'm currently using zephyr 3.5.0.

A good discussion of the modifying zephyr to build into ram was here:
https://github.com/zephyrproject-rtos/zephyr/discussions/36800
(with important updates from May 2023 and later)

I also found important tips for the loading process on the NXP forums here:
https://community.nxp.com/t5/i-MX-RT/mimxrt1024-EVK-flash-and-boot/m-p/1306050
(it a 2-step process: use sdphost to download flash loader   and blhost to download application to itcm)

@DerekSnell has been an important contributor in these discussions, thanks you!

For my initial testing I'm working on MIMXRT1062-EVKB.  
Following the discussion in the first link, I modified my custom board dts to place the application into itcm/dtcm.
	chosen {
		//zephyr,flash-controller = &is25wp064;
		//zephyr,code-partition = &slot0_partition;
		zephyr,flash = &itcm;
		zephyr,sram = &dtcm;
        };

I see my code running when I perform west debug.

However when I look into build/zephyr/zephyr.bin   I don't find any isr vectory, and most importantly there isn't a reset vectory

After loading I verfied this:

(.venv) C:\Users\antho>blhost -u 0x15A2,0x0073 -- read-memory 0x0000000 8
Reading memory  [####################################]  100%
00 00 00 00 00 00 00 00
Response status = 0 (0x0) Success.
Response word 1 = 8 (0x8)
Read 8 of 8 bytes.

What am I missing to properly build zephyr.bin?

**Comment by DerekSnell** [comment]
Hi @asteriskSF ,

You can check the zephyr.map file from the linker, generated with the zephyr.bin file.  The vector table is located at `_vector_table`, and the MAP file will give you this address.  In ARM's vector table, the reset vector is at offset 0x4 of the table, the second 32-bit word.  The first word is the initial stack pointer.

Best regards

**Comment by asteriskSF** [comment]
For some reason the vector table is being placed at 0x2000
 *(SORT_BY_ALIGNMENT(.exc_vector_table))
 *(SORT_BY_ALIGNMENT(.exc_vector_table.*))
 .exc_vector_table._vector_table_section
                0x0000000000002000       0x40 zephyr/arch/arch/arm/core/cortex_m/libarch__arm__core__cortex_m.a(vector_table.S.obj)
                0x0000000000002000                _vector_table
 *(SORT_BY_ALIGNMENT(.vectors))
                0x00000000000022c0                _vector_end = .
                0x0000000000002040                . = ALIGN (0x4


Does it need to be at address 0x0 for correct operation? 
Is it ok to have it placed somwhere else as long as I set the jump_address to whatever the contents of _vector_table  + 0x04 show?

**Comment by DerekSnell** [comment, marked as answer]
Hi @asteriskSF ,
The vector table does not need to be at 0x0, although you could place it there.  It will also work at 0x2000.

When the ROM bootloader boots the app, like booting from flash, the boot ROM needs to be told where the vector table is.  But since you are loading to RAM, you will be booting the app manually using the bootloader tools.  Like using the blhost `execute` command that you found at this [post](https://community.nxp.com/t5/i-MX-RT/mimxrt1024-EVK-flash-and-boot/m-p/1306050).

The vector table location is also configured in the ARM core's VTOR register, which does have an alignment requirement.  I don't recall what alignment is required for the M7 core (you can review the VTOR documentation to confirm), but I know the offset of 0x2000 works, as would 0x0.  

The reason the Zephyr app places the vector table at offset 0x2000 is because of the setup for flash boot.  The boot ROM for the RT1xxx devices needs the IVT and FCB at specific offsets in the flash, and the flash image header used for Zephyr defaults to placing the vector table at offset 0x2000.  Since you are not booting from flash, these are not required.  But I suggest you prove out your app executing first with the vector table at 0x2000, before you modify the image further.

Best regards

**Reply by asteriskSF** [reply]
Thanks @DerekSnell I have been able to load my image and show its working.  I can probably live with the reset vector existing at 0x2000 but just in case we decide to change where do I find the flash image header?

Another small question..

The value in the reset vector is 0x3e55  .. seems unusual to me that it is an odd number and not aligned to 4.  in the zephyr.map file I see the __start (e.g. z_arm_reset) symbol is 0x3e54.  Why is the value in the vector table off-by-1 ?

                0x0000000000003e54                __start
                0x0000000000003e54                z_arm_reset

```
00001ff0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00002000: 00a5 0020 553e 0000 b184 0000 293e 0000  ... U>......)>..
00002010: 293e 0000 293e 0000 293e 0000 0000 0000  )>..)>..)>......
00002020: 0000 0000 0000 0000 0000 0000 493f 0000  ............I?..
00002030: 293e 0000 0000 0000 e53e 0000 f14e 0000  )>.......>...N.
```

**Reply by DerekSnell** [reply]
Hi @asteriskSF ,
You can also use the .MAP file to find the source file with the flash image header info.  Just look at the address in the .MAP file, and it will show you the source file and static symbol located there.

As for the address in the reset vector, in the ARM Cortex-M cores, the LSB of the PC indicates if the instruction is in Thumb mode.  For more details, see this [post](https://community.nxp.com/t5/i-MX-RT/Extracting-ResetISR-address-from-bin-file/m-p/1012381).

Best regards

**Reply by asteriskSF** [reply]
Thanks you have been very helpful!

**Comment by tq-delimayuki** [comment]
Hi I still ask myself why it seems to not be possible to declare:
zephyr,itcm = &itcm;
zephyr,dtcm = &dtcm;

and why it is necessary to make this confusing statement: 
zephyr,flash = &itcm;
zephyr,sram = &dtcm;


would be nice to have an explicit answer on that manner, thx!
